### PR TITLE
Update/Replace GPT-3.5-turbo with GPT-4o-mini

### LINKED ISSUE #12724: Update `gpt-3.5-turbo` references to `gpt-4o-mini`
### Summary

OpenAI recently released [GPT-4o mini](https://openai.com/index/gpt-4o-mini-advancing-cost-efficient-intelligence/) model. It is cheaper and more performant than GPT-3.5 Turbo and they are planning to deprecate the GPT-3.5 Turbo in favor of the new one.

We still have [many references](https://github.com/search?q=repo%3Amlflow%2Fmlflow+gpt-3.5-turbo&type=code) to the old model in the documentation, examples, and tutorials. We want to replace those reference to the new `gpt-4o-mini` model.

### Notes

- Setting up dev environment following [the contribution guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#setting-up-the-repository).
- Make sure to open a PR from a **non-master** branch.
- Sign off the commit using the `-s` flag when making a commit:

  ```sh
  git commit -s -m "..."
           # ^^ make sure to use this
  ```

- Run `pre-commit run --all-files` 
- Include `#{issue_number}` (e.g. `#123`) in the PR description when opening a PR.

### PR DESCRIPTION
**Pull request by Acksout**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/Acksout/mlflow/pull/12746?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12746/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12746
```

</p>
</details>

Resolve #12724

This commit is the continuation to complete the migration from GPT-3.5-turbo to GPT-4o-mini across the codebase.

Changes have been made in the following directories:
docs/
examples/
mlflow/
tests/

Excluded directories:
mlflow/examples/llms/RAG/_cache.json
mlflow/examples/llms/RAG/mlflow_docs_scraped.csv

Few mentions of GPT-3.5 still remain but I believe those are necessary for context.

### What changes are proposed in this pull request?

- Update references from gpt-3.5-turbo to gpt-4o-mini in documentation
- Update examples and tutorials to use the new gpt-4o-mini model
- Maintain specific references to GPT-3.5 variants where necessary for backwards compatibility or historical context

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (manual review of updated documentation and examples)

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [x] API references
  - [x] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Description: Update documentation and examples to reflect the use of GPT-4o-mini instead of GPT-3.5-turbo.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none`
- [ ] `rn/breaking-change`
- [ ] `rn/feature`
- [ ] `rn/bug-fix`
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by B-Step62** [comment]
@Acksout I think references in `tests/` should be updated as well. The JSON/CSV file in the examples can be unchanged.

**Comment by Acksout** [comment]
I've made the changes in /test directory. @B-Step62

**Review by B-Step62 — APPROVED** [review]
LGTM, thank you so much for the update @Acksout! I will merge the PR once all the CI tests pass.

File: tests/llama_index/test_llama_index_tracer.py [hunk 1]
```diff
@@ -40,10 +40,10 @@ def _get_all_traces() -> List[Trace]:
 
 @pytest.mark.parametrize("is_async", [True, False])
 def test_trace_llm_complete(is_async):
-    # By default llama-index uses "gpt-3.5-turbo" model that only has chat interface,
+    # By default llama-index uses "gpt-4o-mini" model that only has chat interface,
     # and llama-index redirects completion call to chat endpoint. We use non-chat
     # model here to test trace for completion.
-    model_name = "gpt-3.5-turbo-instruct"
+    model_name = "gpt-4o-mini"
```
Comment (B-Step62) [inline tests/llama_index/test_llama_index_tracer.py:46]: We cannot replace this and L71 below for the reason in the comment.

**Comment by B-Step62** [comment]
There are a few failures in cross version tests:

1. `langchain / autolog` - It seems the OpenAI class in LangChain still uses `gpt-3.5-turbo-instruct` by default. We can pass the new model name to the constructor [here](https://github.com/Acksout/mlflow/blob/2b2e3a7c5d48cc44ce52bc090b5e6d1c9ca76a00/tests/langchain/test_langchain_autolog.py#L65C11-L65C17), or leave the assertion the old model name.
2. `llama-index / models / 0.10.44` - Let's revert the change in file `tests/llama_index/test_llama_index_tracer.py`. We have to use `gpt-3.5-turbo-instruct` to test completion behavior.
3. `llama-index / models / 0.10.57` - The failures are due to unrelated breaking change in the library. We are fixing it so please ignore this one🙂

**Comment by Acksout** [comment]
Mentioned breaking changes were reversed.

File: tests/langchain/test_langchain_autolog.py [hunk 2]
```diff
@@ -296,7 +296,7 @@ def test_llmchain_autolog():
         assert spans[1].inputs == ["What is a good name for a company that makes MLflow?"]
         assert spans[1].outputs["generations"][0][0]["text"] == "test"
         attrs = spans[1].attributes
-        assert attrs["invocation_params"]["model_name"] == "gpt-3.5-turbo-instruct"
+        assert attrs["invocation_params"]["model_name"] == "gpt-3.5-turbo"
```
Comment (B-Step62) [inline tests/langchain/test_langchain_autolog.py:299]:
```suggestion
        assert attrs["invocation_params"]["model_name"] == "gpt-3.5-turbo-instruct"
```

File: tests/llama_index/test_llama_index_tracer.py [hunk 3]
```diff
@@ -43,7 +43,7 @@ def test_trace_llm_complete(is_async):
     # By default llama-index uses "gpt-3.5-turbo" model that only has chat interface,
     # and llama-index redirects completion call to chat endpoint. We use non-chat
     # model here to test trace for completion.
-    model_name = "gpt-3.5-turbo-instruct"
+    model_name = "gpt-3.5-turbo"
```
Comment (B-Step62) [inline tests/llama_index/test_llama_index_tracer.py:46]:
```suggestion
    model_name = "gpt-3.5-turbo-instruct"
```

File: tests/llama_index/test_llama_index_tracer.py [hunk 4]
```diff
@@ -68,7 +68,7 @@ def test_trace_llm_complete(is_async):
 
 
 def test_trace_llm_complete_stream():
-    model_name = "gpt-3.5-turbo-instruct"
+    model_name = "gpt-3.5-turbo"
```
Comment (B-Step62) [inline tests/llama_index/test_llama_index_tracer.py:71]:
```suggestion
    model_name = "gpt-3.5-turbo-instruct"
```

**Comment by B-Step62** [comment]
Merged the PR with some adjustment to pass tests. Thank you so much for the effort @Acksout!!🙂

**Comment by Acksout** [comment]
Thank you for merging the PR and making the necessary adjustments! @B-Step62 🙂

**Pull request merged**